knotx-core/knotx-server/src/main/java/com/cognifide/knotx/server/KnotxServerVerticle.java

```diff
@@ -121,15 +123,22 @@ private JsonObject createEngineRequest(RepositoryResponse repositoryResponse, Ht
         return new TemplateEngineRequest(
                 repositoryResponse.getData(),
                 request.method(),
-                getPreservedHeaders(request))
+                getPreservedHeaders(request.headers()))
                 .toJsonObject();
     }
 
-    private MultiMap getPreservedHeaders(HttpServerRequest request) {
+    private void rewriteHeaders(HttpServerRequest httpServerRequest, MultiMap headers) {
+        if (headers != null) {
```

Comment:
And here you don't need to check for null headers - request will always return MultiMap even if no headers in actual requests